### ISSUE TITLE
is there a any way to convert raw volume data to nii file?

### ISSUE THREAD

**Issue by HyunjiEllenPak** [jackyko1991/vnet-tensorflow#5]
I am trying to use your code with my customized CT volume raw data.

you used simpleITK module to read the nii file. but it doesn't support raw file .
so I'm considering two options.
One is that I change your code to read raw file and the second option is that I change my raw file into nii file which you used.
I think second option is much better! but I didn't find the way how to convert raw file into nii file. :(

Do you think is there any way?

**Comment by jackyko1991** [comment]
which type of raw data you are using? DICOM or other format?

Hyunji Pak <notifications@github.com> 於 2019年3月19日週二 上午10:25寫道：

**Comment by HyunjiEllenPak** [comment]
![image](https://user-images.githubusercontent.com/16814077/54580623-5ffe9300-4a4c-11e9-8e71-b6bbf1fad182.png)
It is ".raw file" which includes big-endian,16bit data !

**Comment by jackyko1991** [comment]
DICOM is a typical medical image format to distribute medical image data.

Contact the CT supporting staffs to ask them help you to the conversion. This type of raw data is vendor specific and they must have a manual to export data as dicom.

**Comment by HyunjiEllenPak** [comment]
yes, I changed my raw data into nifty file. thanks!

and now I'm training with my data but I'm struggling to train. Because the prediction from the training is so bad :( .maybe something is wrong in my process.

the object that I should segment is very small comparing to the entire volume.  It's like the object is 1500-times than background in volume.
so in this case, which one should I adjust to get high accuracy with your VNet? Could you give me some advices? 

In this case, the network should extract and combine the deep hierachical features to infer the very small object. so one V-Net in your 'train.py' is not enough ?

**Comment by jackyko1991** [comment]
There are several parameters you should take care of:
1. background to label volume ratio, usually this two value has huge difference
2. dropout ratio, that means to discard some patch data that does not contain labels

here are some recommendations:
1. set min_pixel to a higher value at the beginning of training, then decrease it when training converges
2. set drop ratio to 0 at early training stage, then steadily increase it
3. carefully adjust patch_size, patch_layer and resample size (`NiftiDataset.Resample((0.4,0.4,0.4)),`), this is especailly essential for good training result

choose a good preprocessing pipeline will also enhance training efficiency, e.g., usage a manual normalization instead of auto image normalization for CT images

**Comment by HyunjiEllenPak** [comment]
@jackyko1991 
thank you, I'm trying to do! :)

At that time you implemented the mandible segmentation, the size of VNet in your github is enough to get high accuracy?

**Comment by HyunjiEllenPak** [comment]
@jackyko1991 HI, leave answers, please .

**Comment by jackyko1991** [comment]
@HyunjiEllenPak 
1. I recommend you to use a faster learning rate, I believe 1e-2 will be a good start. I have trained many segmentation models with this network with both 1060 and titan Xp. 2 hours is definitely a very short training time even for Titan. I have used this network for teeth segmentation, the case is even more difficult as the enamel, dentine and pupl cavity are in different CBCT density with significant metallic artifacts. This takes me 3 weeks to get the final training result. But for most cases, an overnight training would be sufficient.
You should first take a look at the sorensen loss and see if it has become plateau. If it is still decreasing that means you can use a faster learning speed.

2. 1 and 2 are the same for binary detection, just 2 will occupy a little bit more memory. I have asked the author on this issue https://github.com/MiguelMonteiro/VNet-Tensorflow/issues/1

**Comment by HyunjiEllenPak** [comment]
@jackyko1991 
Thank you for the advice.

![image](https://user-images.githubusercontent.com/16814077/54750924-ab679b80-4c1c-11e9-9d21-d14714b18505.png)

I trained the mandible segmentation for about 10 hours using Jaccard loss with only 1 training data to check if the model could reach the plateau(=overfitting statue for the training data). I thought that if the model was in overfitting with training data, the model could infer the similar result as the ground truth for the training data. 
But the result was not like what I meant. It still doesn't segment the trabecular bone which has dark intensity. 

Anyway, as you said the VNet is feasible to train the mandible, I'll try to find the fine model.

**Comment by jackyko1991** [comment]
@HyunjiEllenPak I usually use sorensen loss as suggested by the original paper, with learning rate start from 1e-2 then manually perform staircase learning rate decay when loss is plateau.

For convenience you could first try with lower resolution, i.e. 
```python
trainTransforms = [
                NiftiDataset.Normalization(),
                NiftiDataset.Resample(0.4356), #try to use higher value here, e.g. 3
                NiftiDataset.Padding((FLAGS.patch_size, FLAGS.patch_size, FLAGS.patch_layer)),
                NiftiDataset.RandomCrop((FLAGS.patch_size, FLAGS.patch_size, FLAGS.patch_layer),FLAGS.drop_ratio,FLAGS.min_pixel),
                NiftiDataset.RandomNoise()
                ]
```

**Comment by HyunjiEllenPak** [comment]
@jackyko1991 
With the params that you recommended, I got the fine result of mandible segmentation after I trained the model during 12hours. The loss graph is clear. Thank you. 👍 
![image](https://user-images.githubusercontent.com/16814077/54798868-ee1d8800-4c9d-11e9-832d-732cd2fced8a.png)
and I'll refine the result, tuning parameters more and adding the post-processing. According to the research, The CRF could refine the segmentation result well, did you try this before?

**Comment by jackyko1991** [comment]
@HyunjiEllenPak CRF do give a 1-2% improvement after training. But it oftens take more training time than the VNet as they only run on CPU.

I have tried two toolkits: https://github.com/mbickel/DenseInferenceWrapper and https://github.com/jackyko1991/Bayesian-Image-Label-Correction

For the Bayesian model, I recommend you to run it in 2D mode with parallel threads, or first crop image to the mandible region before running the code

**Comment by HyunjiEllenPak** [comment]
@jackyko1991 Thank you. I already found the first one toolkit before and the second one is new.

and actually, I would like to segment the much smaller one which is nerve, than the mandible in CT. 
After I segmented the mandible and then tried to segment the nerve in the mandible region. (It's a cascaded method). But it seems that I failed to segment the nerve region. Instead of the nerve, the mandible was segmented again. 
The nerve region is 82-times than background in crop CT, unlikely to mandible which was just 12-times than the background. I think the nerve was processed like noise or same to the background in the network(in my case, VNet). so the network couldn't recognize the nerve. 

Have you ever experienced segmentation which is about a much smaller object than the background? what do you think about my opinion? I'd like to take some direction.

**Comment by jackyko1991** [comment]
@HyunjiEllenPak actually I have exactly same research before on mandible nerve detection. The problem is very challenging as the nerve canal move in a direction that is very hard to be visualized in the orthogonal planes.

I am currently working in extending the vnet with an attention mechanism and hopefully this can help to do improve the detection rate of small regions.

What I suggest is to convert the binary label to a probability map using Gaussian blur/ distance conversion, then minimize the l2 loss with softmax output of vnet with this map.

Another method is to perform landmark detection, which I believe ResNet can do a more effecient job than Vnet on this issue

**Comment by HyunjiEllenPak** [comment]
@jackyko1991 
Thank you, I'll try your suggestions.
Actually, I had tried the first suggestion with 2.5d-UNet before, but I didn't get any good result.
and the second suggestion,  did you mean 3d-landmark detection using 3d-volume or 2d-image?
I knew there are some landmark detections in 2d-image such as facial landmark detection, but I don't know about landmark detection in 3d-volume, even whether it is feasible or not now.
